src/components/ChannelList/utils.ts

```diff
@@ -30,3 +30,133 @@ export const moveChannelUp = <
+  channels: Array<Channel<SCG>>;
+  channelToMove: Channel<SCG>;
+  sort: ChannelSort<SCG>;
+  /**
+   * If the index of the channel within `channels` list which is being moved upwards
+   * (`channelToMove`) is known, you can supply it to skip extra calculation.
+   */
+  channelToMoveIndexWithinChannels?: number;
+};
+
+/**
+ * This function should not be used to move pinned already channels.
+ */
+export const moveChannelUpwards = <
+  SCG extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+>({
+  channels,
+  channelToMove,
+  channelToMoveIndexWithinChannels,
+  sort,
+}: MoveChannelUpwardsParams<SCG>) => {
+  // get index of channel to move up
+  const targetChannelIndex =
+    channelToMoveIndexWithinChannels ??
+    channels.findIndex((channel) => channel.cid === channelToMove.cid);
+
+  const targetChannelExistsWithinList = targetChannelIndex >= 0;
+  const targetChannelAlreadyAtTheTop = targetChannelIndex === 0;
+
+  // pinned channels should not move within the list based on recent activity, channels which
+  // receive messages and are not pinned should move upwards but only under the last pinned channel
+  // in the list
+  const considerPinnedChannels = shouldConsiderPinnedChannels(sort);
+
+  if (targetChannelAlreadyAtTheTop) return channels;
+
+  const newChannels = [...channels];
+
+  // target channel index is known, remove it from the list
+  if (targetChannelExistsWithinList) {
+    newChannels.splice(targetChannelIndex, 1);
+  }
+
+  // as position of pinned channels has to stay unchanged, we need to
+  // find last pinned channel in the list to move the target channel after
+  let lastPinnedChannelIndex: number | null = null;
+  if (considerPinnedChannels) {
+    lastPinnedChannelIndex = findLastPinnedChannelIndex({ channels: newChannels });
+  }
+
+  // re-insert it at the new place (to specific index if pinned channels are considered)
+  newChannels.splice(
+    typeof lastPinnedChannelIndex === 'number' ? lastPinnedChannelIndex + 1 : 0,
+    0,
+    channelToMove,
+  );
+
+  return newChannels;
+};
+
+/**
+ * Returns true only if `{ pinned_at: -1 }` or `{ pinned_at: 1 }` option is first within the `sort` array.
+ */
+export const shouldConsiderPinnedChannels = <SCG extends ExtendableGenerics>(
+  sort: ChannelListProps<SCG>['sort'],
+) => {
+  if (!sort) return false;
+
+  if (!Array.isArray(sort)) return false;
+
+  const [option] = sort;
+
+  if (!option?.pinned_at) return false;
+
+  return Math.abs(option.pinned_at) === 1;
+};
+
+/**
+ * Returns `true` only if `archived` property is set to `false` within `filters`.
+ */
+export const shouldConsiderArchivedChannels = <SCG extends ExtendableGenerics>(
+  filters: ChannelListProps<SCG>['filters'],
+) => {
+  if (!filters) return false;
+
+  return !filters.archived;
+};
+
+export const isChannelPinned = <SCG extends ExtendableGenerics>(channel: Channel<SCG>) => {
+  if (!channel) return false;
+
+  const member = channel.state.membership;
```

Comment:
Hey!

I'm working on implementing pinning channels on our app.
I would just like to inform you that the `channel.state.membership` does not get updated when user pins/unpins a channel when a `member.updated` websocket notifications is received. Those this will not represent the "current" value of the `pinned_at` state.

Reply:
Hey, @michalvankodev, I know - I have another branch with appropriate change ready for a PR, it will be linked to this one and new version of the client will be required before releasing this feature. :)

Reply:
This one: https://github.com/GetStream/stream-chat-js/pull/1409